vault/external_key_registry.go

```diff
@@ -0,0 +1,628 @@
+// Key config entries. There are a few here, but the general idea is that a
+// config must be either inherited or "typed", and we can move from an inherited
+// config to a typed config, but not the other way.
+func (r *ExternalKeyRegistry) validateConfigTransition(
+	typeBefore bool, // Was "type" set before the change?
+	typeAfter bool, // Is "type" set after the change?
+	inheritsBefore bool, // Was "inherits" set before the change?
+	inheritsAfter bool, // Is "inherits" set after the change?
+) error {
+	switch {
+	case (typeBefore && inheritsBefore) || (typeAfter && inheritsAfter):
+		return fmt.Errorf(`the "type" and "inherits" fields are mutually exclusive`)
+	case !typeAfter && !inheritsAfter:
+		return fmt.Errorf(`either the "type" or the "inherits" field must be set`)
+
+	case typeBefore && !typeAfter:
+		return fmt.Errorf(`cannot remove field "type"`)
+
+	case inheritsBefore && !inheritsAfter && !typeAfter:
+		return fmt.Errorf(`removing field "inherits" requires adding field "type"`)
+	case inheritsBefore && inheritsAfter && typeAfter:
+		return fmt.Errorf(`adding field "type" requires removing field "inherits"`)
+
+	default:
+		return nil
+	}
+}
+
+// DELETE /sys/external-keys/configs/:config-name
+func (r *ExternalKeyRegistry) DeleteConfig(
+	ctx context.Context, req *logical.Request, data *framework.FieldData,
+) (*logical.Response, error) {
+	name := data.Get("config").(string)
+	storage := r.storageViewForRequest(req)
+
+	r.storageLock.Lock()
+	defer r.storageLock.Unlock()
+
+	keysView := logical.NewStorageView(storage, r.keyListPath(name))
+	if err := logical.ClearViewWithLogging(ctx, keysView, r.logger); err != nil {
+		return handleError(err)
+	}
+
+	if err := storage.Delete(ctx, r.configPath(name)); err != nil {
+		return handleError(err)
+	}
+
+	return nil, nil
+}
+
+// LIST /sys/external-keys/configs/:config-name/keys
+func (r *ExternalKeyRegistry) ListKeys(
+	ctx context.Context, req *logical.Request, data *framework.FieldData,
+) (*logical.Response, error) {
+	name := data.Get("config").(string)
+	storage := r.storageViewForRequest(req)
+
+	keys, err := storage.List(ctx, r.keyListPath(name))
+	if err != nil {
+		return handleError(err)
+	}
+
+	return logical.ListResponse(keys), nil
+}
+
+// GET /sys/external-keys/configs/:config-name/keys/:key-name
+func (r *ExternalKeyRegistry) GetKey(
+	ctx context.Context, req *logical.Request, data *framework.FieldData,
+) (*logical.Response, error) {
+	name := path.Join(data.Get("config").(string), data.Get("key").(string))
+	storage := r.storageViewForRequest(req)
+
+	key, err := r.getKeyCommon(ctx, storage, name)
+	switch {
+	case err != nil:
+		return handleError(err)
+	case key == nil:
+		return nil, logical.CodedError(http.StatusNotFound, "key not found")
+	}
+
+	resp := &logical.Response{Data: make(map[string]any)}
+	for k, v := range key.Values {
+		resp.Data[k] = v
+	}
+
+	return resp, nil
+}
+
+// PUT /sys/external-keys/configs/:config-name/keys/:key-name
```

Comment:
(ah, so I was correct above; `keys` are nested under configs. Given we want locally-unique key names, shouldn't we address that? :thinking:)